RecoHI/HiTracking/plugins/HIMultiTrackSelector.cc

```diff
@@ -0,0 +1,708 @@
+  evt.getByToken( src_, hSrcTrack );
+  const TrackCollection& srcTracks(*hSrcTrack);
+  assert(mvaVals_.size()==srcTracks.size());
+
+ // get hits in track..
+  Handle<TrackingRecHitCollection> hSrcHits;
+  evt.getByToken( hSrc_, hSrcHits );
+  const TrackingRecHitCollection & srcHits(*hSrcHits);
+
+
+  auto_ptr<edm::ValueMap<float> >mvaValValueMap = auto_ptr<edm::ValueMap<float> >(new edm::ValueMap<float>);
+  edm::ValueMap<float>::Filler mvaFiller(*mvaValValueMap);
+
+
+  if(!useAnyMVA_){
+    // mvaVals_ already initalized...
+    mvaFiller.insert(hSrcTrack,mvaVals_.begin(),mvaVals_.end());
+    mvaFiller.fill();
+    evt.put(mvaValValueMap,"MVAVals");
+    return;
+  }
+
+  bool checkvertex = std::find(mvavars_indices.begin(), mvavars_indices.end(),dxyperdxyerror)!=mvavars_indices.end()
+        || std::find(mvavars_indices.begin(), mvavars_indices.end(),dzperdzerror)!=mvavars_indices.end();
+
+  size_t current = 0;
+  for (TrackCollection::const_iterator it = srcTracks.begin(), ed = srcTracks.end(); it != ed; ++it, ++current) {
+    const Track & trk = * it;
+
+    float mvavalues[15];
+    mvavalues[ndof] = trk.ndof();
+    mvavalues[nlayers] = trk.hitPattern().trackerLayersWithMeasurement();
+    mvavalues[nlayers3d] =  trk.hitPattern().pixelLayersWithMeasurement() + trk.hitPattern().numberOfValidStripLayersWithMonoAndStereo();
+    mvavalues[nlayerslost] =  trk.hitPattern().trackerLayersWithoutMeasurement(reco::HitPattern::TRACK_HITS);
+    mvavalues[chi2n_no1dmod] = trk.normalizedChi2();
+    mvavalues[chi2perdofperlayer] = mvavalues[chi2n_no1dmod]/mvavalues[nlayers];
+
+
+    float chi2n1d =  trk.normalizedChi2();
+    int count1dhits = 0;
+    auto ith = trk.extra()->firstRecHit();
+    auto  edh = ith + trk.recHitsSize();
+    for (; ith<edh; ++ith) {
+      const TrackingRecHit & hit = srcHits[ith];
+      if (hit.dimension()==1) ++count1dhits;
+    }
+    if (count1dhits > 0) {
+      float chi2 = trk.chi2();
+      float ndof = trk.ndof();
+      chi2n1d = (chi2+count1dhits)/float(ndof+count1dhits);
+    }
+
+    mvavalues[chi2n] = chi2n1d;//chi2 and 1d modes
+
+    mvavalues[eta] = trk.eta();
+    mvavalues[relpterr] = float(trk.ptError())/std::max(float(trk.pt()),0.000001f);
+    mvavalues[nhits] = trk.numberOfValidHits();
+
+    int lostIn = trk.hitPattern().numberOfLostTrackerHits(reco::HitPattern::MISSING_INNER_HITS);
+    int lostOut = trk.hitPattern().numberOfLostTrackerHits(reco::HitPattern::MISSING_OUTER_HITS);
+    mvavalues[minlost] = std::min(lostIn,lostOut);
+    mvavalues[lostmidfrac] = trk.numberOfLostHits() / (trk.numberOfValidHits() + trk.numberOfLostHits());
+
+    mvavalues[etaerror] = trk.etaError();
+
+    float reldz = 0;
+    float reldxy = 0;
+    if (checkvertex) {
+      int vtxind = 0; // only first vertex is taken into account for the speed purposes
+      float dxy = trk.dxy(vertices[vtxind].position()), dxyE =  sqrt(trk.dxyError()*trk.dxyError()+vertices[vtxind].xError()*vertices[vtxind].yError());
```

Comment:
dxyE is mildly incorrect (OK for symmetric vertices): the vertex error component should be added in the direction orthogonal to pt